pkg/operators/combiner/combiner.go

```diff
@@ -0,0 +1,301 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package combiner is a data operator that combines data from a same data
+// source coming from different targets into a single data source. This is
+// useful when we run the same gadget in a distributed environment like
+// Kubernetes and we want to combine the data from all the nodes on the client
+// side so that we can perform further operations on the combined data, e.g.
+// sorting. Notice that this operator is useful only when we have data sources
+// of type array.
+package combiner
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+const (
+	// We need it to be the first operator to run in client side. Only operators
+	// registering data sources should run before the combiner operator.
+	Priority         = -500
+	DataSourcePrefix = "combined"
+	OperatorName     = "Combiner"
+)
+
+type combinerOperator struct{}
+
+func (o *combinerOperator) Name() string {
+	return OperatorName
+}
+
+func (o *combinerOperator) Init(params *params.Params) error {
+	return nil
+}
+
+func (o *combinerOperator) GlobalParams() api.Params {
+	return nil
+}
+
+func (o *combinerOperator) InstanceParams() api.Params {
+	return nil
+}
+
+func (o *combinerOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, paramValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	targetsVar, ok := gadgetCtx.GetVar(runtime.NumRunTargets)
+	if !ok {
+		gadgetCtx.Logger().Debugf("No targets found. Skipping combiner operator")
+		return nil, nil
+	}
+
+	targets, ok := targetsVar.(int)
+	if !ok {
+		return nil, fmt.Errorf("invalid type for number of targets: %T, expected \"int\"", targetsVar)
+	}
+
+	if targets == 0 {
+		return nil, nil
+	}
+
+	for _, ds := range gadgetCtx.GetDataSources() {
+		if ds.Type() == datasource.TypeArray {
+			gadgetCtx.Logger().Debugf("Array data source found (%s). Activating combiner operator", ds.Name())
+			return &combinerOperatorInstance{
+				targets:    targets,
+				pktBuffers: make(map[datasource.DataSource]chan datasource.PacketArray),
+			}, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (o *combinerOperator) Priority() int {
+	return Priority
+}
+
+type combinerOperatorInstance struct {
+	// Number of targets to (ideally) wait for before emitting the combined data
+	targets int
+
+	// Map of packet buffers to send data to the combiner data source
+	pktBuffers map[datasource.DataSource]chan datasource.PacketArray
+
+	done chan struct{}
+}
+
+func (o *combinerOperatorInstance) Name() string {
+	return OperatorName
```

Comment:
Do we want the same name for the instance or another one (e.g. _instance is added)

Reply:
Yep: https://github.com/inspektor-gadget/inspektor-gadget/pull/3068#discussion_r1697657792